### PR TITLE
JAMES-2596 Add listSources API to RecipientRewriteTable

### PR DESCRIPTION
**Pull request by hoangdat**
JIRA: https://issues.apache.org/jira/browse/JAMES-2596

I'm not sure about this, please guide me when I misunderstand this ticket

### CODE REVIEW

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 1]
```diff
@@ -96,7 +97,8 @@ public void setMappingLimit(int mappingLimit) throws IllegalArgumentException {
 
     @Override
     public Mappings getMappings(String user, Domain domain) throws ErrorMappingException, RecipientRewriteTableException {
-        return getMappings(User.fromLocalPartWithDomain(user, domain), mappingLimit);
+        Mappings mappings = getMappings(User.fromLocalPartWithDomain(user, domain), mappingLimit);
+        return mappings;
```
Comment (chibenwa) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: Why this change?

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 2]
```diff
@@ -297,6 +299,11 @@ public void removeGroupMapping(MappingSource source, String address) throws Reci
         removeMapping(source, mapping);
     }
 
+    @Override
+    public List<MappingSource> listSources(Mapping mapping) throws RecipientRewriteTableException {
+        return null;
```
Comment (chibenwa) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: I'm pretty sure it is not what is expected... Maybe we should filter the all Mappings content here...

File: server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractJamesUsersRepository.java [hunk 3]
```diff
@@ -288,4 +291,9 @@ public void addGroupMapping(MappingSource source, String address) throws Recipie
     public void removeGroupMapping(MappingSource source, String address) throws RecipientRewriteTableException {
         throw new RecipientRewriteTableException("Read-Only RecipientRewriteTable");
     }
+
+    @Override
+    public List<MappingSource> listSources(Mapping mapping) throws RecipientRewriteTableException {
+        return ImmutableList.of();
```
Comment (chibenwa) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractJamesUsersRepository.java]: By having `listSources` implemented as a default method we would avoid repeating ourselves.

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java [hunk 4]
```diff
@@ -354,4 +354,32 @@ public void removeGroupMappingShouldDelete() throws ErrorMappingException, Recip
         assertThat(virtualUserTable.getMappings(user, domain))
             .isEqualTo(MappingsImpl.empty());
     }
+
+    @Test
+    public void listSourcesShouldReturnWhenHasMapping() throws RecipientRewriteTableException {
+        String user = "test";
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java]: user should be a constant, address too .

Having these variables makes test reading harder...

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java [hunk 5]
```diff
@@ -354,4 +354,32 @@ public void removeGroupMappingShouldDelete() throws ErrorMappingException, Recip
...
+        MappingSource source = MappingSource.fromUser(user, domain);
+        MappingSource source2 = MappingSource.fromDomain(Domain.LOCALHOST);
+        Mapping mapping = Mapping.group(address);
+
+        virtualUserTable.addMapping(source, mapping);
+        virtualUserTable.addMapping(source2, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source, source2);
+    }
 }
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java]: You are missing...

 - Tests with forward + address that should be handled...
 - Tests that a mapping without source will lead to an empty list being returned.

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java (hunk 2 above)
Comment (hoangdat) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: IMO, why we should filter all Mappings content instead of using query in each of implementation?

Comment (chibenwa) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: > why we should filter all Mappings content instead of using query in each of implementation?

I do not fully understand your question, but the goal would be to provide a default, correct, but not efficient behaviour by default.

This enables:

 - Implementers of RecipientRewriteTable to not have their implementation broken because of a missing method in the implementation
 - Allows working on the webadmin API without having to wait for an implementation to implement this API.

**Comment by hoangdat** [comment]
Build triggered. sha1 is merged.

**Comment by hoangdat** [comment]
Build started sha1 is merged.

**Comment by hoangdat** [comment]
Build finished. No test results found.

**Comment by hoangdat** [comment]
Build triggered. sha1 is merged.

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java (hunk 1 above)
Comment (hoangdat) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: Srr, I made it for debug, Removed.

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java (hunk 5 above)
Comment (hoangdat) [reply server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java]: Add more tests

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java [hunk 6]
```diff
@@ -106,6 +109,31 @@ public String toString() {
      */
     Map<MappingSource, Mappings> getAllMappings() throws RecipientRewriteTableException;
 
+    default List<MappingSource> listSources(Mapping mapping) throws RecipientRewriteTableException {
+        if (!isSupportedListSources(mapping)) {
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: We should rather throw to make explicit the fact that it is not supported, no? (Preconditions)

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java [hunk 7]
```diff
@@ -106,6 +109,31 @@ public String toString() {
...
+        return getAllMappings().entrySet().stream()
+            .filter(entry -> filterMapping(entry, mapping))
+            .map(Map.Entry::getKey)
+            .collect(ImmutableList.toImmutableList());
+    }
+
+    default boolean filterMapping(Map.Entry<MappingSource, Mappings> entry, Mapping mapping) {
+        return entry.getValue()
+            .asStream()
+            .anyMatch(map -> map.equals(mapping));
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: `.anyMatch(mapping::equals)` ?

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java [hunk 8]
```diff
@@ -106,6 +109,31 @@ public String toString() {
...
+    }
+
+    default boolean isSupportedListSources(Mapping mapping) {
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: s/isSupportedListSources/supportsSourceListing/

File: server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractJamesUsersRepository.java [hunk 9]
```diff
@@ -21,7 +21,9 @@
 
 import java.util.HashMap;
 import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
```
Comment (chibenwa) [inline server/data/data-library/src/main/java/org/apache/james/user/lib/AbstractJamesUsersRepository.java]: I bet these imports are not used

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java [hunk 10]
```diff
@@ -267,91 +262,136 @@ public void testAliasDomainMapping() throws ErrorMappingException, RecipientRewr
...
+    @Test
+    public void listSourcesShouldReturnEmptyWhenHasRegexMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.LOCALHOST;
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        Mapping mapping = Mapping.regex("regex");
+
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).isEmpty();
+    }
```
Comment (chibenwa) [inline server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java]: What happens when we are listing sources for an Error mapping? a domain mapping? A regex mapping?

What happens when we are looking for a mapping that do not exist (and thus have no sources?)

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java [hunk 11]
```diff
@@ -112,26 +113,27 @@ public String toString() {
...
-            .filter(entry -> filterMapping(entry, mapping))
+            .filter(entry -> entryContainsMapping(entry, mapping))
             .map(Map.Entry::getKey)
-            .collect(ImmutableList.toImmutableList());
+            .collect(Guavate.toImmutableList());
     }
 
-    default boolean filterMapping(Map.Entry<MappingSource, Mappings> entry, Mapping mapping) {
+    default boolean entryContainsMapping(Map.Entry<MappingSource, Mappings> entry, Mapping mapping) {
         return entry.getValue()
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: `entry.getValue().contains(mapping);`

Comment (chibenwa) [reply server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: BTW IMO we don't need this method at all it makes code reading harder

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java [hunk 12]
```diff
@@ -112,26 +113,27 @@ public String toString() {
...
-        .of(Mapping.Type.Group, Mapping.Type.Forward, Mapping.Type.Address);
+    EnumSet<Mapping.Type> listSourcesSupportedType = EnumSet.of(
+        Mapping.Type.Group,
+        Mapping.Type.Forward,
+        Mapping.Type.Address);
 
     default boolean supportsSourceListing(Mapping mapping) {
-        return listSourcesSupportedType.stream()
-            .anyMatch(mapping.getType()::equals);
+        return listSourcesSupportedType.contains(mapping.getType());
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/RecipientRewriteTable.java]: This method can be inlined too without hurting readability, no?

**Comment by chibenwa** [comment]
Merged
